src/parser/parse.js

```diff
@@ -16,16 +17,41 @@ import parseWDL from './WDL/parse';
  * @param {string} text - Text file contents to parse.
  * @param {object} [opts={}] - Parser options.
  * @param {string} [opts.format='wdl'] - Workflow definition format ('wdl', 'cwl').
- * @returns {ParseResult} Parsing result object
+ * @returns {Promise} Parsing result object
  */
 function parse(text, opts = {}) {
   const format = opts.format || 'wdl';
+  const baseURI = opts.baseURI;
   if (format === 'wdl') {
-    return parseWDL(text, opts);
+    if (opts.zipFile) {
+      return JSZip.loadAsync(opts.zipFile).then((files) => {
+        const zipWdlFiles = [];
+
+        files.forEach((relativePath, zipEntry) => {
+          if (!zipEntry.dir && zipEntry.name.indexOf('.wdl') === zipEntry.name.length - 4) {
+            zipWdlFiles.push(zipEntry);
+          }
+        });
+
+        return Promise.all(zipWdlFiles.map(zipWdlFile => zipWdlFile.async('string').then(str => ({
+          name: zipWdlFile.name.split('/').pop(),
+          wdl: str,
+        }))))
+          .then(wdlArray => parseWDL(text, { wdlArray, baseURI }));
+      }, (e) => {
+        throw new Error(`Parse zip file: ${e}`);
+      });
+    }
+    return new Promise((resolve) => {
+      parseWDL(text, opts).then((data) => {
+        resolve(data);
+      });
+    });
```

Comment:
- Am I right saying that with this implementation we are unable to catch exceptions raised in the `parseWDL()` function?
- I'm not sure that we need a `new Promise()` here since we already have one from `parseWDL()` function.

Please test thoroughly if a user can successfully intercept all exceptions by artificially generating them in different parts of code, and then adjust the code correspondingly.

Reply:
Fixed